### PR TITLE
[feature-wip](Cloud) Introduce obj storage client interface to recycler

### PR DESCRIPTION
**Pull request by ByteYue**
## Proposed changes

This is one subsequent pr of #35307 which does the same thing to recycler to extract basic interface to suite different kinds of Object Storage.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by ByteYue** [comment]
run buildall

File: cloud/src/recycler/obj_store_accessor.h [hunk 1]
```diff
@@ -69,4 +71,44 @@ class ObjStoreAccessor {
...
+
+struct ObjectCompleteMultiParts {};
+
+struct ObjectStorageResponse {
+    ObjectStorageResponse(int r) : ret(r) {}
+    int ret {0};
+};
+
+// wrapper class owned by concret fs
+class ObjStorageClient {
```
Comment (gavinchou) [inline cloud/src/recycler/obj_store_accessor.h:94]: Add enough/detailed comments to describe the behavior of each function as much as possible

File: cloud/src/recycler/obj_store_accessor.h [hunk 2]
```diff
@@ -69,4 +71,44 @@ class ObjStoreAccessor {
...
+struct ObjectStorageDeleteExpiredOptions {
+    ObjectStoragePathOptions path_opts;
+    std::function<std::string(const std::string& path)> relative_path_factory;
+};
+
+struct ObjectCompleteMultiParts {};
+
+struct ObjectStorageResponse {
+    ObjectStorageResponse(int r) : ret(r) {}
+    int ret {0};
```
Comment (gavinchou) [inline cloud/src/recycler/obj_store_accessor.h:90]: Is this for "return code"?
If yes, try to make it `enum class` or pre-define some constants for codes

File: cloud/src/recycler/s3_obj_client.cpp [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+#include <aws/s3/model/ListObjectsV2Request.h>
+#include <aws/s3/model/PutObjectRequest.h>
+
+#include "common/logging.h"
+#include "common/sync_point.h"
+
+namespace doris::cloud {
+
+#ifndef UNIT_TEST
+#define HELP_MACRO(ret, req, point_name)
```
Comment (gavinchou) [inline cloud/src/recycler/s3_obj_client.cpp:35]: typo, HELPER_MACRO

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

File: cloud/src/recycler/s3_obj_client.cpp [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+                    .tag("bucket", opts.bucket)
+                    .tag("prefix", opts.prefix)
+                    .tag("responseCode", static_cast<int>(outcome.GetError().GetResponseCode()))
+                    .tag("error", outcome.GetError().GetMessage());
+            return -1;
+        }
+        const auto& result = outcome.GetResult();
+        VLOG_DEBUG << "get " << result.GetContents().size() << " objects";
+        for (const auto& obj : result.GetContents()) {
+            files->push_back({obj.GetKey().substr(opts.prefix.size() + 1), obj.GetSize(),
```
Comment (platoneko) [inline cloud/src/recycler/s3_obj_client.cpp:116]: 在 `ObjStoreAccessor` 中 `ObjectMeta` 的 `path` 需要存除去 Accessor prefix 的那段，而不是除去传入的 prefix 的段。e.g. accessor.prefix = "doris_s3_vault", opts.preifx = "doris_s3_vault/data/114115/"

File: be/src/io/fs/s3_obj_storage_client.cpp
Comment (platoneko) [inline be/src/io/fs/s3_obj_storage_client.cpp:1]: 在 BE 中 FileInfo 中的 file_name 不应该包含 list 使用的 prefix 那段 e.g.:
prefix = "doris_s3_vault/data/114115/", key = "doris_s3_vault/data/114115/abc_0.dat", file_name = "abc_0.dat"
prefix = "doris_s3_vault/data/100/114115/", key = "doris_s3_vault/data/100/114115/abc/0.dat", file_name = "abc/0.dat"

**Review by platoneko — CHANGES_REQUESTED** [review]

**Review by gavinchou — APPROVED** [review]

**Review by gavinchou — APPROVED** [review]

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Review by platoneko — APPROVED** [review]

**Pull request merged**